### PR TITLE
feat: DBTP-1071 Generate terraform config for environment pipeline

### PR DESCRIPTION
**Pull request by a-gleeson**
Addresses [DBTP-1071](https://uktrade.atlassian.net/browse/DBTP-1071)

Updated the `platform-helper pipeline generate` command. To unblock [DBTP-1073](https://uktrade.atlassian.net/jira/software/c/projects/DBTP/boards/458?label=SMALL&selectedIssue=DBTP-1073).

- This command originally generated the copilot pipelines and we have extended the functionality to generate the terraform main.tf file for the environment pipelines for terraform projects.
- You can override the version of `terraform-platform-modules` the environment pipelines are configured to use with the flag `--terraform-platform-modules-version`, otherwise the default release version is used. 

Output of running `platform-helper pipeline generate` command:
Screenshot 2024-10-28 at 18 56 07

### Still Todo
- [ ] Separate PR to update documentation repo (now in progress)
- [x] add the autogenerated terraform folder  to `.gitignore` in `demodjango-deploy`. https://github.com/uktrade/demodjango-deploy/pull/228

---
### Title:
- [ ] Scope included as per [conventional commits](https://www.conventionalcommits.org/en/v1.0.0/)
- [ ] Ticket reference included (unless it's a quick out of ticket thing)
### Description:
- [ ] Link to ticket included (unless it's a quick out of ticket thing)
- [ ] Includes tests (or an explanation for why it doesn't)
- [ ] If the work includes user interface changes, before and after screenshots included in description
- [ ] Includes any applicable changes to the documentation in this code base
- [ ] Includes link(s) to any applicable changes to the documentation in the [DBT Platform Documentation](https://platform.readme.trade.gov.uk/) (can be to a pull request)
### Tasks:
- [ ] [Trigger the pull request regression tests for this branch](https://github.com/uktrade/platform-tools?tab=readme-ov-file#regression-tests) and confirm that they are passing

[DBTP-1071]: https://uktrade.atlassian.net/browse/DBTP-1071?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

[DBTP-1073]: https://uktrade.atlassian.net/browse/DBTP-1073?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dbt_platform_helper/commands/pipeline.py [hunk 1]
```diff
@@ -33,15 +35,22 @@ def pipeline():
...
     """Given a platform-config.yml file, generate environment and service
     deployment pipelines."""
     pipeline_config = load_and_validate_platform_config()
 
     no_codebase_pipelines = CODEBASE_PIPELINES_KEY not in pipeline_config
     no_environment_pipelines = ENVIRONMENTS_KEY not in pipeline_config
+    no_terraform_environment_pipelines = ENVIRONMENT_PIPELINES_KEY not in pipeline_config
 
-    if no_codebase_pipelines and no_environment_pipelines:
+    if no_codebase_pipelines and no_environment_pipelines and no_terraform_environment_pipelines:
```
Comment (WillGibson) [inline dbt_platform_helper/commands/pipeline.py:53]:
```suggestion
    if not has codebase_pipelines and not has_legacy_environment_pipelines and not has_environment_pipelines:
```

File: dbt_platform_helper/commands/pipeline.py [hunk 2]
```diff
@@ -33,15 +35,22 @@ def pipeline():
...
+    default=DEFAULT_TERRAFORM_PLATFORM_MODULES_VERSION,
+)
+def generate(terraform_platform_modules_version):
     """Given a platform-config.yml file, generate environment and service
     deployment pipelines."""
     pipeline_config = load_and_validate_platform_config()
 
     no_codebase_pipelines = CODEBASE_PIPELINES_KEY not in pipeline_config
     no_environment_pipelines = ENVIRONMENTS_KEY not in pipeline_config
+    no_terraform_environment_pipelines = ENVIRONMENT_PIPELINES_KEY not in pipeline_config
```
Comment (WillGibson) [inline dbt_platform_helper/commands/pipeline.py:51]: I would be inclined to reverse the extra word as we'll soon be losing all the legacy stuff. Perhaps actually reverse the boolean too, then use it in all the other places we perform a the same check...
```suggestion
    has_codebase_pipelines = CODEBASE_PIPELINES_KEY in pipeline_config
    has_legacy_environment_pipelines = ENVIRONMENTS_KEY in pipeline_config
    has_environment_pipelines = ENVIRONMENT_PIPELINES_KEY in pipeline_config
```

File: dbt_platform_helper/commands/pipeline.py [hunk 3]
```diff
@@ -57,18 +66,28 @@ def generate():
         abort_with_error(f'There is no CodeStar Connection named "{app_name}" to use')
 
     base_path = Path(".")
-    pipelines_dir = base_path / f"copilot/pipelines"
+    copilot_pipelines_dir = base_path / f"copilot/pipelines"
+
+    _clean_pipeline_config(copilot_pipelines_dir)
 
-    _clean_pipeline_config(pipelines_dir)
+    if is_terraform_project() and ENVIRONMENT_PIPELINES_KEY in pipeline_config:
```
Comment (WillGibson) [inline dbt_platform_helper/commands/pipeline.py:73]:
```suggestion
    if is_terraform_project() and has_environment_pipelines:
```

File: dbt_platform_helper/commands/pipeline.py [hunk 4]
```diff
@@ -57,18 +66,28 @@ def generate():
...
+        environment_pipelines = pipeline_config[ENVIRONMENT_PIPELINES_KEY]
 
+        for config in environment_pipelines.values():
+            aws_account = config.get("account")
+            _generate_terraform_environment_pipeline_manifest(
+                pipeline_config["application"],
+                aws_account,
+                terraform_platform_modules_version,
+            )
     if not is_terraform_project() and ENVIRONMENTS_KEY in pipeline_config:
```
Comment (WillGibson) [inline dbt_platform_helper/commands/pipeline.py:83]:
```suggestion
    if not is_terraform_project() and has_legacy_pipelines:
```

File: dbt_platform_helper/templates/environment-pipelines/main.tf [hunk 5]
```diff
@@ -0,0 +1,53 @@
+# {% extra_header %}
+# {% version_info %}
+locals {
+  platform_config    = yamldecode(file("../../platform-config.yml"))
+  all_pipelines      = local.platform_config["environment_pipelines"]
+  pipelines = { for pipeline, config in local.platform_config["environment_pipelines"] : pipeline => config if config.account == "{{ aws_account }}" }
```
Comment (WillGibson) [inline dbt_platform_helper/templates/environment-pipelines/main.tf:6]:
```suggestion
  pipelines          = { for pipeline, config in local.platform_config["environment_pipelines"] : pipeline => config if config.account == "{{ aws_account }}" }
```

File: dbt_platform_helper/templates/environment-pipelines/main.tf [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+      source  = "hashicorp/aws"
+      version = "~> 5"
+    }
+  }
+}
+
+
+module "environment-pipelines" {
+  source = "git::https://github.com/uktrade/terraform-platform-modules.git//environment-pipelines?depth=1&ref={{ terraform_platform_modules_version }}"
+  #  source = "../../../terraform-platform-modules/environment-pipelines"
```
Comment (WillGibson) [inline dbt_platform_helper/templates/environment-pipelines/main.tf:38]: I'm not sure we should keep this in the autogenerated file...
```suggestion
```

**Review by WillGibson — COMMENTED** [review]
Looking good. Some fairly trivial suggestions.

File: dbt_platform_helper/commands/pipeline.py (hunk 1 above)
Comment (a-gleeson) [reply dbt_platform_helper/commands/pipeline.py:53]: Good catch!

File: dbt_platform_helper/commands/pipeline.py (hunk 2 above)
Comment (a-gleeson) [reply dbt_platform_helper/commands/pipeline.py:51]: Thanks, I am inclined to agree and I'll make the update.

File: dbt_platform_helper/templates/environment-pipelines/main.tf (hunk 6 above)
Comment (a-gleeson) [reply dbt_platform_helper/templates/environment-pipelines/main.tf]: is it maybe useful to leave for people to use? Or maybe in the case of 1073, we would add a command for this command to generate the main.tf with a realitive source instead? Just thinking out loud here

Comment (WillGibson) [reply dbt_platform_helper/templates/environment-pipelines/main.tf]: I don't really mind, it's not a deal breaker, but it feels untidy to leave it in our production code.

**Review by WillGibson — APPROVED** [review]

**Comment by JohnStainsby** [comment]
Nice 👍

What do you think about also checking the default version of `terraform-platform-modules` set in platform-config.yml? The [environment generate command](https://github.com/uktrade/platform-tools/blob/main/dbt_platform_helper/commands/environment.py#L352) checks for CLI override, then platform-config, then the default.

```yml
application: demodjango

default_versions:
  platform-helper: 11.0.1
  terraform-platform-modules: 5.5.2
```

**Review by JohnStainsby — APPROVED** [review]

File: dbt_platform_helper/commands/pipeline.py [hunk 7]
```diff
@@ -170,3 +204,44 @@ def _create_file_from_template(
...
+        }
+    )
+
+    dir_path = f"terraform/environment-pipelines/{aws_account}"
+    makedirs(dir_path, exist_ok=True)
+
+    click.echo(mkfile(".", f"{dir_path}/main.tf", contents, overwrite=True))
+
+
+def _determine_terraform_platform_modules_version(
```
Comment (a-gleeson) [inline dbt_platform_helper/commands/pipeline.py:237]: Could this be extracted to utlis? I think this is repeated in the environment command too

**Pull request merged**